text/0227-separate-semantic-conventions.md

```diff
@@ -0,0 +1,179 @@
+
+These exceptions exist because:
+
+- Stable portions of the specification already rely on these conventions
+- These conventions are required to implement an SDK today.
+
+As such, the Specification should define the absolute minimum of reserved or
+required attribute names and their interaction with the SDK.
+
+## Internal details
+
+The following process would be used to ensure semantic conventions are
+seamlessly moved to their new location. This process lists steps in order:
+
+- A moratorium will be placed on Semantic Convention PRs to the specififcation
+  repository. (Caveat that PRs related to this proposal would be allowed).
+- Interactions between Semantic Conventions and the Specification will be
+  extracted such that the Specification can place requirements on Semantic
+  Conventions and *normative* specification language will remain in the
+  core specification directories.
+- A new repository `open-telemetry/semantic-conventions` will be constructed with
+  the proposed format and necessary `Makefile` / tooling.
+  - The new repository would be created by using `git filter-branch` to preserve
+    all existing semantic convention history. *This means all existing
+    semantic conventions will be in the new repository*.
+  - Github Actions, `Makefile` tooling and contributing / readmes would be
+    updated for the separate repository.
+  - **Note: At this point, the new location for semantic conventions should
+      be adoptable/usable.**
+- Semantic conventions in the Specification will be marked as deprecated with
+  links to the new location.
+  - The semconv YAML files in the specification repository *will be deleted*.
+  - All semconv markdown files will be updated such that:
+    - They no longer generate from YAML files.
+    - They include a header denoting deprecation and move to the new repository.
+- Instrumentation authors will update their code generation to pull from the new
+  semconv repository instead of the specification repository..
+- The new repository will have its first released versioned be `2.0.0`, to avoid
+    confusion with specification version.
+
+## Trade-offs and mitigations
+
+This proposal has a few drawbacks:
+
+- The semantic conventions will no longer be easily referencable form the specification.
+  - This is actually a benefit. We can ensure isolation of convention from
+    specification and require the Specification to use firm language for
+    attributes it requires, like `service.name`.
+  - We will provide links from existing location to the new location.
+- Semantic Convention version will no longer match the specification version.
+  - Instrumentation authors will need to consume a separate semantic-convention
+    bundle from Specification bundle. What used to be ONE upgrade effort will
+    now be split into two (hopefully smaller) efforts.
```

Comment:
In the java SIG, we discussed breaking out a new repository to hold the generated code for the semantic conventions. Something like `semantic-conventions-java`. This artifact could be versioned in sync with the `semantic-conventions` repo and the other java repo `opentelemetry-java` and `opentelemetry-java-instrumentation` would take a dependency on it.

Reply:
We have this in JS as well. In practice it is tough to version exactly with the semconv and we decided to keep our own version numbers. Now that it's a bit more stable and we're not refactoring it and making changes like that it might be easier.

Reply:
I would recommend at least using a bifurcated version strategy here:

1. The version of semconv
2. The version of the *codegen* for semconv.


This way if codegen changes in a breaking manner you can bump versions.  Otherwise you'd be forced to limit codegen fixes, and you may find you *need* to make breaking changes there.  The semconv version won't be able to reflect that.

Note: this is the same debate as OTLP versioning and generated protobuf repositories.